osutil/strace/timing.go

```diff
@@ -0,0 +1,213 @@
+func NewExecveTiming(nSlowestSamples int) *ExecveTiming {
+	return &ExecveTiming{nSlowestSamples: nSlowestSamples}
+}
+
+func (stt *ExecveTiming) addExeRuntime(exe string, totalSec float64) {
+	stt.exeRuntimes = append(stt.exeRuntimes, ExeRuntime{
+		Exe:      exe,
+		TotalSec: totalSec,
+	})
+	stt.prune()
+}
+
+// prune() ensures the number of exeRuntimes stays with the nSlowestSamples
+// limit
+func (stt *ExecveTiming) prune() {
+	for len(stt.exeRuntimes) > stt.nSlowestSamples {
+		fastest := 0
+		for idx, rt := range stt.exeRuntimes {
+			if rt.TotalSec < stt.exeRuntimes[fastest].TotalSec {
+				fastest = idx
+			}
+		}
+		// delete fastest element
+		stt.exeRuntimes = append(stt.exeRuntimes[:fastest], stt.exeRuntimes[fastest+1:]...)
+	}
+}
+
+func (stt *ExecveTiming) Display(w io.Writer) {
+	if len(stt.exeRuntimes) == 0 {
+		return
+	}
+	fmt.Fprintf(w, "Slowest %d exec calls during snap run:\n", len(stt.exeRuntimes))
+	for _, rt := range stt.exeRuntimes {
+		fmt.Fprintf(w, "  %2.3fs %s\n", rt.TotalSec, rt.Exe)
+	}
+	fmt.Fprintf(w, "Total time: %2.3fs\n", stt.TotalTime)
+}
+
+type exeStart struct {
+	start float64
+	exe   string
+}
+
+type pidTracker struct {
+	pidToExeStart map[string]exeStart
+}
+
+func newPidTracker() *pidTracker {
+	return &pidTracker{
+		pidToExeStart: make(map[string]exeStart),
+	}
+
+}
+
+func (pt *pidTracker) Get(pid string) (startTime float64, exe string) {
+	if exeStart, ok := pt.pidToExeStart[pid]; ok {
+		return exeStart.start, exeStart.exe
+	}
+	return 0, ""
+}
+
+func (pt *pidTracker) Add(pid string, startTime float64, exe string) {
+	pt.pidToExeStart[pid] = exeStart{start: startTime, exe: exe}
+}
+
+func (pt *pidTracker) Del(pid string) {
+	delete(pt.pidToExeStart, pid)
+}
+
+// lines look like:
+// PID   TIME              SYSCALL
+// 17363 1542815326.700248 execve("/snap/brave/44/usr/bin/update-mime-database", ["update-mime-database", "/home/egon/snap/brave/44/.local/"...], 0x1566008 /* 69 vars */) = 0
+var execveRE = regexp.MustCompile(`([0-9]+)\ +([0-9.]+) execve\(\"([^"]+)\"`)
+
+// lines look like:
+// PID   TIME              SYSCALL
+// 14157 1542875582.816782 execveat(3, "", ["snap-update-ns", "--from-snap-confine", "test-snapd-tools"], 0x7ffce7dd6160 /* 0 vars */, AT_EMPTY_PATH) = 0
+var execveatRE = regexp.MustCompile(`([0-9]+)\ +([0-9.]+) execveat\(.*\["([^"]+)"`)
+
+// lines look like (both SIGTERM and SIGCHLD need to be handled):
+// PID   TIME                  SIGNAL
+// 17559 1542815330.242750 --- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=17643, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
+var sigChldTermRE = regexp.MustCompile(`[0-9]+\ +([0-9.]+).*SIG(CHLD|TERM)\ {.*si_pid=([0-9]+),`)
+
+// all lines start with this:
+// PID   TIME
+// 21616 1542882400.198907 ....
+var timeRE = regexp.MustCompile(`[0-9]+\ +([0-9.]+).*`)
+
+func handleExecMatch(trace *ExecveTiming, pt *pidTracker, match []string) error {
+	if len(match) == 0 {
+		return nil
+	}
+	pid := match[1]
```

Comment:
the pid here is of the invoking process right? not the invoked one?

this bit of code merits some more comments

Reply:
I added some comments now.